File: search/digital-search-tree/tree.go
package digital

import (
	"algorithm-class/inst"
)

type Tree struct {
	root *Node
	size uint
	bits uint // 가장 큰 비트 번호. 4개의 비트가 있다면 3이 bits가 된다.
}

func NewTree(bits uint) *Tree {
	return &Tree{bits: bits}
}

func (t *Tree) Size() uint { return t.size }

func (t *Tree) Bits() uint { return t.bits }

func (t *Tree) Insert(ic *inst.Counter, value int) {
	t.root = insert(ic, t.root, value, t.bits)
	t.size++
}

func (t *Tree) Search(ic *inst.Counter, value int) bool {
	return search(ic, t.root, value, t.bits)
}

type Node struct {
	Value int
	Left  *Node
	Right *Node
}

func search(ic *inst.Counter, r *Node, value int, bits uint) bool {
	x := r
	for x != nil {
		ic.Once(inst.Indirect)
		if x.Value == value {
			return true
		}
		ic.Once(inst.Indirect)
		if lookBit(value, bits) {
			ic.Once(inst.Indirect)
			x = x.Right
		} else {
			ic.Once(inst.Indirect)
			x = x.Left
		}
		bits--
	}
	return false
}

func insert(ic *inst.Counter, r *Node, value int, bits uint) *Node {
	rejectCounter(ic)

	if r == nil {
		return &Node{
			Value: value,
		}
	}
	if r.Value == value {
		return r
	}
	if lookBit(value, bits) {
		r.Right = insert(ic, r.Right, value, bits-1)
		return r
	} else {
		r.Left = insert(ic, r.Left, value, bits-1)
		return r
	}
}

func lookBit(n int, i uint) bool {
	return (n>>i)&1 == 1
}

func rejectCounter(ic *inst.Counter) {
	if ic != nil {
		panic(" is not yet implemented")
	}
}

File: search/digital-search-tree/tree.rs
const MAXBITS: usize = 4;

#[derive(Debug)]
struct Node {
    value: usize,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

impl Node {
    fn new(value: usize) -> Node {
        Node {
            value,
            left: None,
            right: None,
        }
    }

    fn insert(&mut self, value: usize) {
        if bits(value, 0, 1) == 1 {
            if let Some(ref mut right) = self.right {
                right.insert(value >> 1);
            } else {
                self.right = Some(Box::new(Node::new(1)));
            }
            return;
        }
        if let Some(ref mut left) = self.left {
            left.insert(value >> 1);
        } else {
            self.left = Some(Box::new(Node::new(0)));
        }
    }

    fn search(&self, value: usize) -> bool {
        if bits(value, 0, 1) == 1 {
            return match self.right {
                Some(ref right) => right.search(value >> 1),
                None => value == 1,
            };
        }
        match self.left {
            Some(ref left) => left.search(value >> 1),
            None => value == 0,
        }
    }
}

fn bits(n: usize, shift: usize, size: usize) -> usize {
    (n >> shift) & !(!0 << size)
}

struct Tree {
    root: Option<Box<Node>>,
}

impl Tree {
    fn new() -> Tree {
        Tree { root: None }
    }

    fn insert(&mut self, value: usize) {
        match self.root {
            Some(ref mut root) => root.insert(value),
            None => self.root = Some(Box::new(Node::new(value))),
        }
    }

    fn search(&self, value: usize) -> bool {
        match self.root {
            Some(ref root) => root.search(value),
            None => false,
        }
    }
}

fn main() {
    let mut tree = Tree::new();
    for i in 0..8 {
        println!("insert {}", i);
        tree.insert(i);
    }
    println!("{:#?}", tree.root);
    for i in 0..10 {
        if tree.search(i) {
            println!("found {}", i);
        } else {
            println!("not found {}", i);
        }
    }
}
